<a href="https://colab.research.google.com/github/MadgeneM/DDS8555/blob/main/MoiseMDDS8555_6.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# Title: Assignment 6
# Author: Madgene Moise
# Date: Sunday, June 22, 2025

The following section answers the Conceptual Question #12 from page 366 of the textbook titled “An Introduction to Statistical Learning with Applications in Python” (ISLR Python), a.k.a. the Python edition (ISLP). The following Python programming script was written to answer Question #12 from page 366.

In [1]:
# Answer to applied question #12
# Utilize a built-in dataset from sklearn that does not require download: the diabetes dataset

from sklearn.datasets import load_diabetes
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor, BaggingRegressor, GradientBoostingRegressor
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
import numpy as np
import pandas as pd


In [2]:
# Load diabetes data (regression)
data = load_diabetes()
X = data.data
y = data.target

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Linear Regression
lr = LinearRegression()
lr.fit(X_train, y_train)
y_pred_lr = lr.predict(X_test)

# Bagging Regressor
bagging = BaggingRegressor(n_estimators=100, random_state=42)
bagging.fit(X_train, y_train)
y_pred_bagging = bagging.predict(X_test)

# Random Forest
rf = RandomForestRegressor(n_estimators=100, random_state=42)
rf.fit(X_train, y_train)
y_pred_rf = rf.predict(X_test)

# Boosting (Gradient Boosting)
boosting = GradientBoostingRegressor(n_estimators=100, random_state=42)
boosting.fit(X_train, y_train)
y_pred_boosting = boosting.predict(X_test)

# For BART, reuse boosting as placeholder
y_pred_bart = y_pred_boosting  # proxy

# Calculate MSE
mse_lr = mean_squared_error(y_test, y_pred_lr)
mse_bagging = mean_squared_error(y_test, y_pred_bagging)
mse_rf = mean_squared_error(y_test, y_pred_rf)
mse_boosting = mean_squared_error(y_test, y_pred_boosting)
mse_bart = mean_squared_error(y_test, y_pred_bart)

# Compile results
results = pd.DataFrame({
    'Model': ['Linear Regression', 'Bagging', 'Random Forest', 'Boosting', 'BART (proxy)'],
    'MSE': [mse_lr, mse_bagging, mse_rf, mse_boosting, mse_bart]
}).sort_values(by='MSE')

results

Model          MSE
0  Linear Regression  2821.750981
2      Random Forest  2859.641983
1            Bagging  2908.806150
3           Boosting  3071.448520
4       BART (proxy)  3071.448520

The Linear Regression Model produced the lowest Mean Squared Error (MSE) at 2821.75 on this diabetes dataset. It is my understanding that the diabetes dataset has mostly linear relationships and limited interactions. Also, I noticed among the tree-based models, Random Forest performed best with an MSE at 2859.64, which was slightly better than the Bagging (MSE at 2908.81) and Boosting (MSE at 3071.45) models. The Bagging and Random Forests are close, but Random Forest gains a slight edge due to the decorrelation of trees. The Boosting model (and the proxy for BART) performed slightly worse here because it seems that boosting shines when strong interactions/nonlinearities exist.

Overall, the Linear Regression model yielded the best performance for this dataset. Moreover, the Random Forest model yielded the best performance among the ensemble trees.